boa/src/syntax/parser/expression/assignment/arrow_function.rs

```diff
@@ -0,0 +1,161 @@
+//! Arrow function parsing.
+//!
+//! More information:
+//!  - [MDN documentation][mdn]
+//!  - [ECMAScript specification][spec]
+//!
+//! [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions
+//! [spec]: https://tc39.es/ecma262/#sec-arrow-function-definitions
+
+use super::AssignmentExpression;
+use crate::syntax::{
+    ast::{
+        node::{FormalParameter, Node},
+        punc::Punctuator,
+        token::TokenKind,
+    },
+    parser::{
+        function::{FormalParameters, FunctionBody},
+        AllowAwait, AllowIn, AllowYield, Cursor, ParseError, ParseResult, TokenParser,
+    },
+};
+
+/// Arrow function parsing.
+///
+/// More information:
+///  - [MDN documentation][mdn]
+///  - [ECMAScript specification][spec]
+///
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions
+/// [spec]: https://tc39.es/ecma262/#prod-ArrowFunction
+#[derive(Debug, Clone, Copy)]
+pub(in crate::syntax::parser) struct ArrowFunction {
+    allow_in: AllowIn,
+    allow_yield: AllowYield,
+    allow_await: AllowAwait,
+}
+
+impl ArrowFunction {
+    /// Creates a new `ArrowFunction` parser.
+    pub(in crate::syntax::parser) fn new<I, Y, A>(
+        allow_in: I,
+        allow_yield: Y,
+        allow_await: A,
+    ) -> Self
+    where
+        I: Into<AllowIn>,
+        Y: Into<AllowYield>,
+        A: Into<AllowAwait>,
+    {
+        Self {
+            allow_in: allow_in.into(),
+            allow_yield: allow_yield.into(),
+            allow_await: allow_await.into(),
```

Comment:
How come into() is needed here? (just for me to understand)

Reply:
It's not realy needed, but I used generics in the function signature: `I: Into<AllowIn>` and so on. This allows us to create a parser with booleans too: `ArrowFunction(true, false, true)`.

We could just remove the generics from the `new()` functions and pass the wrapping structures, which would improve compile times, and is probably just as clear / developer friendly.